File: src/components/OneParser/ModbusMapper.ts
/**
 * Modbus Address Mapper
 *
 * Converts LS PLC device addresses to Modbus addresses according to the
 * defined mapping rules. This mapping aligns with the Rust ModServerSync
 * module for consistent address translation.
 */

import type {
  DeviceAddress,
  DeviceType,
  ModbusAddress,
  ModbusAddressType,
  MappingRule,
} from './types';
import { isBitDevice, isWordDevice } from './types';

// ============================================================================
// Default Mapping Rules
// ============================================================================

/**
 * Default Modbus mapping rules for LS PLC devices
 *
 * These mappings align with the Rust ModServerSync module offsets:
 * - Bit devices map to Coils/Discrete Inputs
 * - Word devices map to Holding Registers
 */
export const DEFAULT_MAPPING_RULES: MappingRule[] = [
  // Bit devices to Coils
  { device: 'P', modbusType: 'discrete', offset: 0 },    // Input Relay -> Discrete Input (read from)
  { device: 'M', modbusType: 'coil', offset: 0 },        // Auxiliary Relay -> Coil 0-8191
  { device: 'K', modbusType: 'coil', offset: 8192 },     // Keep Relay -> Coil 8192-10239
  { device: 'T', modbusType: 'coil', offset: 10240 },    // Timer Contact -> Coil 10240-12287 (read-only)
  { device: 'C', modbusType: 'coil', offset: 12288 },    // Counter Contact -> Coil 12288-14335 (read-only)
  { device: 'F', modbusType: 'discrete', offset: 2048 }, // Special Relay -> Discrete Input 2048+ (read-only)

  // Word devices to Holding Registers
  { device: 'D', modbusType: 'holding', offset: 0 },     // Data Register -> Holding 0-9999
  { device: 'R', modbusType: 'holding', offset: 10000 }, // Retentive Data Register -> Holding 10000-19999
  { device: 'Z', modbusType: 'holding', offset: 20000 }, // Index Register -> Holding 20000-20015
  { device: 'N', modbusType: 'holding', offset: 20016 }, // Link Data Register -> Holding 20016+
];

/**
 * Special device addresses for timer/counter current values
 * These are virtual word devices that map to specific Holding Register ranges.
 */
export const SPECIAL_MAPPINGS = {
  /** Timer current value (TD) - Holding Register offset 28208 */
  TD: { modbusType: 'holding' as ModbusAddressType, offset: 28208 },
  /** Counter current value (CD) - Holding Register offset 30256 */
  CD: { modbusType: 'holding' as ModbusAddressType, offset: 30256 },
} as const;

// ============================================================================
// Modbus Mapper Class
// ============================================================================

/**
 * Modbus Address Mapper
 *
 * Provides bidirectional mapping between LS PLC device addresses and
 * Modbus addresses according to configurable mapping rules.
 */
export class ModbusMapper {
  private rules: Map<DeviceType, MappingRule>;

  /**
   * Create a new ModbusMapper
   * @param customRules - Optional custom rules to override defaults
   */
  constructor(customRules?: MappingRule[]) {
    this.rules = new Map();

    // Initialize with default rules
    for (const rule of DEFAULT_MAPPING_RULES) {
      this.rules.set(rule.device, rule);
    }

    // Override with custom rules if provided
    if (customRules) {
      for (const rule of customRules) {
        this.rules.set(rule.device, rule);
      }
    }
  }

  /**
   * Map device address to Modbus address
   *
   * @param deviceAddr - PLC device address to map
   * @returns Modbus address or null if mapping not possible
   *
   * @example
   * ```typescript
   * const mapper = new ModbusMapper();
   * mapper.mapToModbus({ device: 'M', address: 100 });
   * // Returns: { type: 'coil', address: 100 }
   *
   * mapper.mapToModbus({ device: 'K', address: 0 });
   * // Returns: { type: 'coil', address: 8192 }
   *
   * mapper.mapToModbus({ device: 'D', address: 100, bitIndex: 5 });
   * // Returns: { type: 'coil', address: 1605 } (100*16 + 5)
   * ```
   */
  mapToModbus(deviceAddr: DeviceAddress): ModbusAddress | null {
    const rule = this.rules.get(deviceAddr.device);
    if (!rule) {
      return null;
    }

    // Handle indexed addressing - cannot be mapped statically
    if (deviceAddr.indexRegister !== undefined) {
      console.warn(
        `Indexed address ${deviceAddr.device}[Z${deviceAddr.indexRegister}] cannot be mapped statically`
      );
      return null;
    }

    // Handle bit access on word devices
    if (deviceAddr.bitIndex !== undefined && isWordDevice(deviceAddr.device)) {
      // Word device bit access maps to coil area
      // Address = (rule.offset + word_address) * 16 + bit_index
      // This maps D0.0-D0.15 to coils 0-15, D1.0-D1.15 to coils 16-31, etc.
      const wordOffset = rule.offset + deviceAddr.address;
      return {
        type: 'coil',
        address: wordOffset * 16 + deviceAddr.bitIndex,
      };
    }

    return {
      type: rule.modbusType,
      address: rule.offset + deviceAddr.address,
    };
  }

  /**
   * Map Modbus address back to device address
   *
   * Note: This may return multiple possible matches for overlapping ranges.
   * The caller should determine which match is most appropriate based on context.
   *
   * @param modbusAddr - Modbus address to map
   * @returns Array of possible device addresses (may be empty)
   *
   * @example
   * ```typescript
   * const mapper = new ModbusMapper();
   * mapper.mapFromModbus({ type: 'coil', address: 100 });
   * // Returns: [{ device: 'M', address: 100 }]
   *
   * mapper.mapFromModbus({ type: 'coil', address: 8192 });
   * // Returns: [{ device: 'K', address: 0 }, { device: 'M', address: 8192 }]
   * ```
   */
  mapFromModbus(modbusAddr: ModbusAddress): DeviceAddress[] {
    const matches: DeviceAddress[] = [];

    for (const [device, rule] of this.rules) {
      if (rule.modbusType === modbusAddr.type) {
        const deviceAddress = modbusAddr.address - rule.offset;
        if (deviceAddress >= 0) {
          matches.push({
            device,
            address: deviceAddress,
          });
        }
      }
    }

    return matches;
  }

  /**
   * Map timer current value address (TD) to Modbus
   *
   * Timer current values are stored in a separate Holding Register range.
   *
   * @param timerNumber - Timer number (0-2047)
   * @returns Modbus address for timer current value
   *
   * @example
   * ```typescript
   * const mapper = new ModbusMapper();
   * mapper.mapTimerDataToModbus(0);
   * // Returns: { type: 'holding', address: 28208 }
   *
   * mapper.mapTimerDataToModbus(100);
   * // Returns: { type: 'holding', address: 28308 }
   * ```
   */
  mapTimerDataToModbus(timerNumber: number): ModbusAddress {
    return {
      type: 'holding',
      address: SPECIAL_MAPPINGS.TD.offset + timerNumber,
    };
  }

  /**
   * Map counter current value address (CD) to Modbus
   *
   * Counter current values are stored in a separate Holding Register range.
   *
   * @param counterNumber - Counter number (0-2047)
   * @returns Modbus address for counter current value
   *
   * @example
   * ```typescript
   * const mapper = new ModbusMapper();
   * mapper.mapCounterDataToModbus(0);
   * // Returns: { type: 'holding', address: 30256 }
   *
   * mapper.mapCounterDataToModbus(100);
   * // Returns: { type: 'holding', address: 30356 }
   * ```
   */
  mapCounterDataToModbus(counterNumber: number): ModbusAddress {
    return {
      type: 'holding',
      address: SPECIAL_MAPPINGS.CD.offset + counterNumber,
    };
  }

  /**
   * Get the mapping rule for a device type
   *
   * @param device - Device type to get rule for
   * @returns Mapping rule or undefined if not found
   */
  getRule(device: DeviceType): MappingRule | undefined {
    return this.rules.get(device);
  }

  /**
   * Check if a device address is read-only
   *
   * Read-only devices cannot be written to via Modbus:
   * - F (Special Relay) - system status bits
   * - T (Timer Contact) - set by timer logic
   * - C (Counter Contact) - set by counter logic
   *
   * @param deviceAddr - Device address to check
   * @returns True if device is read-only
   */
  isReadOnly(deviceAddr: DeviceAddress): boolean {
    // F (Special Relay), T (Timer Contact), C (Counter Contact) are read-only
    return ['F', 'T', 'C'].includes(deviceAddr.device);
  }

  /**
   * Check if a device is a bit device
   *
   * @param device - Device type to check
   * @returns True if device is a bit device
   */
  isBitDevice(device: DeviceType): boolean {
    return isBitDevice(device);
  }

  /**
   * Check if a device is a word device
   *
   * @param device - Device type to check
   * @returns True if device is a word device
   */
  isWordDevice(device: DeviceType): boolean {
    return isWordDevice(device);
  }
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Format Modbus address for display
 *
 * Uses common prefixes:
 * - C: Coil
 * - DI: Discrete Input
 * - HR: Holding Register
 * - IR: Input Register
 *
 * @param addr - Modbus address to format
 * @returns Formatted string (e.g., "HR:1000", "C:8192")
 *
 * @example
 * ```typescript
 * formatModbusAddress({ type: 'holding', address: 1000 });
 * // Returns: "HR:1000"
 *
 * formatModbusAddress({ type: 'coil', address: 0 });
 * // Returns: "C:0"
 * ```
 */
export function formatModbusAddress(addr: ModbusAddress): string {
  const prefix = {
    coil: 'C',
    discrete: 'DI',
    holding: 'HR',
    input: 'IR',
  }[addr.type];

  return `${prefix}:${addr.address}`;
}

/**
 * Parse Modbus address string
 *
 * Accepts format: PREFIX:ADDRESS
 * - C: Coil
 * - DI: Discrete Input
 * - HR: Holding Register
 * - IR: Input Register
 *
 * @param str - String to parse (e.g., "HR:1000", "C:100")
 * @returns Parsed Modbus address or null if invalid
 *
 * @example
 * ```typescript
 * parseModbusAddress("HR:1000");
 * // Returns: { type: 'holding', address: 1000 }
 *
 * parseModbusAddress("c:100"); // Case insensitive
 * // Returns: { type: 'coil', address: 100 }
 *
 * parseModbusAddress("INVALID");
 * // Returns: null
 * ```
 */
export function parseModbusAddress(str: string): ModbusAddress | null {
  const match = str.match(/^(C|DI|HR|IR):(\d+)$/i);
  if (!match) return null;

  const typeMap: Record<string, ModbusAddressType> = {
    C: 'coil',
    DI: 'discrete',
    HR: 'holding',
    IR: 'input',
  };

  return {
    type: typeMap[match[1].toUpperCase()],
    address: parseInt(match[2], 10),
  };
}

/**
 * Create a default ModbusMapper instance
 *
 * @returns New ModbusMapper with default mapping rules
 */
export function createDefaultMapper(): ModbusMapper {
  return new ModbusMapper();
}

File: src-tauri/src/parser/modbus_mapper.rs
//! Modbus Address Mapper
//!
//! Converts LS PLC device addresses to Modbus addresses according to the
//! defined mapping rules. This mapping aligns with the modserver_sync
//! module for consistent address translation.

use super::types::{BitDeviceType, DeviceAddress, DeviceType, ModbusAddress, ModbusAddressType, WordDeviceType};
use std::collections::HashMap;

// ============================================================================
// Address Mapping Constants (matching modserver_sync.rs)
// ============================================================================

/// M relay → Coil offset (8192 M relays at offset 0)
pub const M_COIL_OFFSET: u16 = 0;
/// K relay → Coil offset (2048 K relays at offset 8192)
pub const K_COIL_OFFSET: u16 = 8192;
/// T contact → Coil offset (2048 T contacts at offset 10240)
pub const T_COIL_OFFSET: u16 = 10240;
/// C contact → Coil offset (2048 C contacts at offset 12288)
pub const C_COIL_OFFSET: u16 = 12288;

/// P relay → Discrete Input offset
pub const P_DISCRETE_OFFSET: u16 = 0;
/// F relay → Discrete Input offset
pub const F_DISCRETE_OFFSET: u16 = 2048;

/// D register → Holding Register offset
pub const D_HR_OFFSET: u16 = 0;
/// R register → Holding Register offset
pub const R_HR_OFFSET: u16 = 10000;
/// Z register → Holding Register offset
pub const Z_HR_OFFSET: u16 = 20000;
/// N register → Holding Register offset
pub const N_HR_OFFSET: u16 = 20016;

/// TD (Timer Current Value) → Holding Register offset
pub const TD_HR_OFFSET: u16 = 28208;
/// CD (Counter Current Value) → Holding Register offset
pub const CD_HR_OFFSET: u16 = 30256;

// ============================================================================
// Mapping Rule Structure
// ============================================================================

/// Device to Modbus mapping rule
#[derive(Debug, Clone)]
pub struct MappingRule {
    /// PLC device type
    pub device: DeviceType,
    /// Modbus memory type to map to
    pub modbus_type: ModbusAddressType,
    /// Address offset for mapping
    pub offset: u16,
}

impl MappingRule {
    /// Get default mapping rules for LS PLC devices (matching modserver_sync.rs)
    pub fn defaults() -> Vec<Self> {
        vec![
            // Bit devices to Coils/Discrete
            MappingRule {
                device: DeviceType::Bit(BitDeviceType::P),
                modbus_type: ModbusAddressType::Discrete,
                offset: P_DISCRETE_OFFSET,
            },
            MappingRule {
                device: DeviceType::Bit(BitDeviceType::M),
                modbus_type: ModbusAddressType::Coil,
                offset: M_COIL_OFFSET,
            },
            MappingRule {
                device: DeviceType::Bit(BitDeviceType::K),
                modbus_type: ModbusAddressType::Coil,
                offset: K_COIL_OFFSET,
            },
            MappingRule {
                device: DeviceType::Bit(BitDeviceType::T),
                modbus_type: ModbusAddressType::Coil,
                offset: T_COIL_OFFSET,
            },
            MappingRule {
                device: DeviceType::Bit(BitDeviceType::C),
                modbus_type: ModbusAddressType::Coil,
                offset: C_COIL_OFFSET,
            },
            MappingRule {
                device: DeviceType::Bit(BitDeviceType::F),
                modbus_type: ModbusAddressType::Discrete,
                offset: F_DISCRETE_OFFSET,
            },
            // Word devices to Holding Registers
            MappingRule {
                device: DeviceType::Word(WordDeviceType::D),
                modbus_type: ModbusAddressType::Holding,
                offset: D_HR_OFFSET,
            },
            MappingRule {
                device: DeviceType::Word(WordDeviceType::R),
                modbus_type: ModbusAddressType::Holding,
                offset: R_HR_OFFSET,
            },
            MappingRule {
                device: DeviceType::Word(WordDeviceType::Z),
                modbus_type: ModbusAddressType::Holding,
                offset: Z_HR_OFFSET,
            },
            MappingRule {
                device: DeviceType::Word(WordDeviceType::N),
                modbus_type: ModbusAddressType::Holding,
                offset: N_HR_OFFSET,
            },
        ]
    }
}

// ============================================================================
// Modbus Mapper
// ============================================================================

/// Modbus Address Mapper
///
/// Provides bidirectional mapping between LS PLC device addresses and
/// Modbus addresses according to configurable mapping rules.
pub struct ModbusMapper {
    rules: HashMap<String, MappingRule>,
}

impl Default for ModbusMapper {
    fn default() -> Self {
        Self::new()
    }
}

impl ModbusMapper {
    /// Create a new ModbusMapper with default rules
    pub fn new() -> Self {
        let mut rules = HashMap::new();
        for rule in MappingRule::defaults() {
            rules.insert(rule.device.as_str().to_string(), rule);
        }
        Self { rules }
    }

    /// Create a ModbusMapper with custom rules
    pub fn with_rules(custom_rules: Vec<MappingRule>) -> Self {
        let mut mapper = Self::new();
        for rule in custom_rules {
            mapper.rules.insert(rule.device.as_str().to_string(), rule);
        }
        mapper
    }

    /// Map device address to Modbus address
    pub fn map_to_modbus(&self, device_addr: &DeviceAddress) -> Option<ModbusAddress> {
        // Cannot map indexed addresses statically
        if device_addr.index_register.is_some() {
            log::warn!(
                "Indexed address {}[Z{}] cannot be mapped statically",
                device_addr.device.as_str(),
                device_addr.index_register.unwrap()
            );
            return None;
        }

        let rule = self.rules.get(device_addr.device.as_str())?;

        // Handle bit access on word devices
        if let Some(bit_index) = device_addr.bit_index {
            if device_addr.device.is_word_device() {
                // Word device bit access maps to coil area
                // Address = (rule.offset + word_address) * 16 + bit_index
                let word_offset = rule.offset as u32 + device_addr.address;
                return Some(ModbusAddress {
                    address_type: ModbusAddressType::Coil,
                    address: (word_offset * 16 + bit_index as u32) as u16,
                });
            }
        }

        Some(ModbusAddress {
            address_type: rule.modbus_type,
            address: rule.offset + device_addr.address as u16,
        })
    }

    /// Map Modbus address back to device address(es)
    ///
    /// Returns multiple matches for overlapping ranges.
    pub fn map_from_modbus(&self, modbus_addr: &ModbusAddress) -> Vec<DeviceAddress> {
        let mut matches = Vec::new();

        for rule in self.rules.values() {
            if rule.modbus_type == modbus_addr.address_type {
                if modbus_addr.address >= rule.offset {
                    let device_address = modbus_addr.address - rule.offset;
                    matches.push(DeviceAddress::new(rule.device, device_address as u32));
                }
            }
        }

        matches
    }

    /// Map timer current value address (TD) to Modbus
    pub fn map_timer_data_to_modbus(&self, timer_number: u16) -> ModbusAddress {
        ModbusAddress {
            address_type: ModbusAddressType::Holding,
            address: TD_HR_OFFSET + timer_number,
        }
    }

    /// Map counter current value address (CD) to Modbus
    pub fn map_counter_data_to_modbus(&self, counter_number: u16) -> ModbusAddress {
        ModbusAddress {
            address_type: ModbusAddressType::Holding,
            address: CD_HR_OFFSET + counter_number,
        }
    }

    /// Check if a device address is read-only
    pub fn is_read_only(&self, device_addr: &DeviceAddress) -> bool {
        match device_addr.device {
            DeviceType::Bit(BitDeviceType::F)
            | DeviceType::Bit(BitDeviceType::T)
            | DeviceType::Bit(BitDeviceType::C) => true,
            _ => false,
        }
    }
}

// ============================================================================
// Utility Functions
// ============================================================================

/// Format Modbus address for display
pub fn format_modbus_address(addr: &ModbusAddress) -> String {
    let prefix = match addr.address_type {
        ModbusAddressType::Coil => "C",
        ModbusAddressType::Discrete => "DI",
        ModbusAddressType::Holding => "HR",
        ModbusAddressType::Input => "IR",
    };
    format!("{}:{}", prefix, addr.address)
}

/// Parse Modbus address string
pub fn parse_modbus_address(s: &str) -> Option<ModbusAddress> {
    let parts: Vec<&str> = s.split(':').collect();
    if parts.len() != 2 {
        return None;
    }

    let address_type = match parts[0].to_uppercase().as_str() {
        "C" => ModbusAddressType::Coil,
        "DI" => ModbusAddressType::Discrete,
        "HR" => ModbusAddressType::Holding,
        "IR" => ModbusAddressType::Input,
        _ => return None,
    };

    let address: u16 = parts[1].parse().ok()?;

    Some(ModbusAddress {
        address_type,
        address,
    })
}

// ============================================================================
// Unit Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_map_m_to_coil() {
        let mapper = ModbusMapper::new();
        let addr = DeviceAddress::bit(BitDeviceType::M, 100);
        let result = mapper.map_to_modbus(&addr);

        assert_eq!(
            result,
            Some(ModbusAddress {
                address_type: ModbusAddressType::Coil,
                address: 100
            })
        );
    }

    #[test]
    fn test_map_k_to_coil_with_offset() {
        let mapper = ModbusMapper::new();
        let addr = DeviceAddress::bit(BitDeviceType::K, 0);
        let result = mapper.map_to_modbus(&addr);

        assert_eq!(
            result,
            Some(ModbusAddress {
                address_type: ModbusAddressType::Coil,
                address: 8192
            })
        );
    }

    #[test]
    fn test_map_d_to_holding() {
        let mapper = ModbusMapper::new();
        let addr = DeviceAddress::word(WordDeviceType::D, 100);
        let result = mapper.map_to_modbus(&addr);

        assert_eq!(
            result,
            Some(ModbusAddress {
                address_type: ModbusAddressType::Holding,
                address: 100
            })
        );
    }

    #[test]
    fn test_map_word_bit_to_coil() {
        let mapper = ModbusMapper::new();
        let addr = DeviceAddress::word_bit(WordDeviceType::D, 1, 5);
        let result = mapper.map_to_modbus(&addr);

        // (0 + 1) * 16 + 5 = 21
        assert_eq!(
            result,
            Some(ModbusAddress {
                address_type: ModbusAddressType::Coil,
                address: 21
            })
        );
    }

    #[test]
    fn test_map_timer_data() {
        let mapper = ModbusMapper::new();
        let result = mapper.map_timer_data_to_modbus(0);

        assert_eq!(
            result,
            ModbusAddress {
                address_type: ModbusAddressType::Holding,
                address: 28208
            }
        );
    }

    #[test]
    fn test_map_counter_data() {
        let mapper = ModbusMapper::new();
        let result = mapper.map_counter_data_to_modbus(100);

        assert_eq!(
            result,
            ModbusAddress {
                address_type: ModbusAddressType::Holding,
                address: 30356
            }
        );
    }

    #[test]
    fn test_format_modbus_address() {
        let addr = ModbusAddress {
            address_type: ModbusAddressType::Holding,
            address: 1000,
        };
        assert_eq!(format_modbus_address(&addr), "HR:1000");
    }

    #[test]
    fn test_parse_modbus_address() {
        let result = parse_modbus_address("HR:1000");
        assert_eq!(
            result,
            Some(ModbusAddress {
                address_type: ModbusAddressType::Holding,
                address: 1000
            })
        );

        assert_eq!(parse_modbus_address("INVALID"), None);
    }

    #[test]
    fn test_is_read_only() {
        let mapper = ModbusMapper::new();
        assert!(mapper.is_read_only(&DeviceAddress::bit(BitDeviceType::F, 0)));
        assert!(mapper.is_read_only(&DeviceAddress::bit(BitDeviceType::T, 0)));
        assert!(mapper.is_read_only(&DeviceAddress::bit(BitDeviceType::C, 0)));
        assert!(!mapper.is_read_only(&DeviceAddress::bit(BitDeviceType::M, 0)));
        assert!(!mapper.is_read_only(&DeviceAddress::word(WordDeviceType::D, 0)));
    }
}
